### PR TITLE
Dedupe query results

### PR DESCRIPTION
**Pull request by stefanholder**
Dedupes the QueryResults of each GPX entry by node id after calling `QueryGraph.lookup` as discussed in #81. This only dedupes tower nodes but not virtual nodes since there is at most one QueryResult per edge and virtual nodes are inserted into the middle of an edge.

Reducing the number of QueryResults improves performance since less shortest/fastest routes need to be computed.

Also updated README.md.

### CODE REVIEW

File: matching-core/src/main/java/com/graphhopper/matching/LocationIndexMatch.java [hunk 1]
```diff
@@ -52,16 +52,23 @@ public LocationIndexMatch(GraphHopperStorage graph, LocationIndexTree index) {
         this.index = index;
     }
 
-    public List<QueryResult> findNClosest(final double queryLat, final double queryLon, final EdgeFilter edgeFilter,
-                                          double gpxAccuracyInMetern) {
+    /**
+     * Returns all edges whose starting points are within the specified radius around the
```
Comment (stefanholder) [inline matching-core/src/main/java/com/graphhopper/matching/LocationIndexMatch.java]: @karussell: Can you please check if this javadoc comment is correct?

Comment (kodonnell) [inline matching-core/src/main/java/com/graphhopper/matching/LocationIndexMatch.java]: I'd reword it to make it clearer that there's no (general) guarantee it finds all such edges (due to searching only nine tiles).

Comment (karussell) [inline matching-core/src/main/java/com/graphhopper/matching/LocationIndexMatch.java]: There is no guarantee if the distance is large than the cell radius which is usually not the case

Comment (stefanholder) [inline matching-core/src/main/java/com/graphhopper/matching/LocationIndexMatch.java]: Instead of writing "Returns all edges whose starting points are within the specified radius ..." can I write "Returns all edges that are within the specified radius ..."?

Comment (karussell) [inline matching-core/src/main/java/com/graphhopper/matching/LocationIndexMatch.java]: Ah, you mean this part. Yes, this version would be better.

File: matching-core/src/main/java/com/graphhopper/matching/MapMatching.java [hunk 2]
```diff
@@ -269,26 +277,46 @@ public MatchResult doWork(List<GPXEntry> gpxList) {
...
+            final List<QueryResult> queryResults = locationIndex.findNClosest(
+                    gpxEntry.lat, gpxEntry.lon, edgeFilter, measurementErrorSigma);
+            gpxEntryLocations.add(queryResults);
         }
         return gpxEntryLocations;
     }
 
+    private List<Collection<QueryResult>> dedupeQueryResultsByClosestNode(
+            List<Collection<QueryResult>> queriesPerEntry) {
+        final List<Collection<QueryResult>> result = new ArrayList<>();
```
Comment (karussell) [inline matching-core/src/main/java/com/graphhopper/matching/MapMatching.java]: Here an init would be good: new ArrayList<>(queriesPerEntry.size())

File: matching-core/src/main/java/com/graphhopper/matching/MapMatching.java [hunk 3]
```diff
@@ -269,26 +277,46 @@ public MatchResult doWork(List<GPXEntry> gpxList) {
...
+
+        final Comparator<QueryResult> dedupeComparator = new Comparator<QueryResult>() {
```
Comment (karussell) [inline matching-core/src/main/java/com/graphhopper/matching/MapMatching.java]: Instead of comparator and TreeSet can we use a `HashMap<closest node, query result>`? (less memory intense as array based)

File: matching-core/src/main/java/com/graphhopper/matching/LocationIndexMatch.java (hunk 1 above)
Comment (stefanholder) [inline matching-core/src/main/java/com/graphhopper/matching/LocationIndexMatch.java]: Done.

File: matching-core/src/main/java/com/graphhopper/matching/MapMatching.java (hunk 2 above)
Comment (stefanholder) [inline matching-core/src/main/java/com/graphhopper/matching/MapMatching.java]: Done.

File: matching-core/src/main/java/com/graphhopper/matching/MapMatching.java (hunk 3 above)
Comment (stefanholder) [inline matching-core/src/main/java/com/graphhopper/matching/MapMatching.java]: Done.

**Comment by stefanholder** [comment]
I added a commit with the changes discussed above. I will squash this with the previous commit when the PR is approved.

File: matching-core/src/main/java/com/graphhopper/matching/MapMatching.java [hunk 4]
```diff
@@ -183,18 +183,29 @@ public MatchResult doWork(List<GPXEntry> gpxList) {
...
             allQueryResults.addAll(qrs);
         queryGraph.lookup(allQueryResults);
 
+        // Different QueryResults can have the same tower node as their closest node.
+        // Hence, we now dedupe the query results of each GPX entry by their closest node.
+        // This must be done after calling queryGraph.lookup() since this replaces some of the
+        // QueryResult nodes with virtual nodes. Virtual nodes are not deduped since there is at
+        // most one QueryResult per edge and virtual nodes are inserted into the middle of an edge.
+        // Reducing the number of QueryResults improves performance since less shortest/fastest
+        // routes need to be computed.
```
Comment (karussell) [inline matching-core/src/main/java/com/graphhopper/matching/MapMatching.java]: Maybe you add a "See #91" here?

Comment (stefanholder) [inline matching-core/src/main/java/com/graphhopper/matching/MapMatching.java]: Since this PR does not contain additional information I don't see the benefit of this. But I will do this if you generally prefer linking to PRs from the code.

Comment (karussell) [inline matching-core/src/main/java/com/graphhopper/matching/MapMatching.java]: Just the number "#91" no link?

Comment (stefanholder) [inline matching-core/src/main/java/com/graphhopper/matching/MapMatching.java]: Done.

Comment (karussell) [inline matching-core/src/main/java/com/graphhopper/matching/MapMatching.java]: Thanks - please feel free to merge or squash (hope this works :)) ... BTW: have now enabled both again as often it is very convenient to avoid manual squashing. But will deactivate it as soon as we see this does not work well somehow

**Comment by stefanholder** [comment]
Squashed commits.

**Comment by stefanholder** [comment]
Referenced PR in commit messages.

**Comment by karussell** [comment]
Thanks - do you want to merge it or prefer that I do it?

**Comment by stefanholder** [comment]
>Thanks - do you want to merge it or prefer that I do it?

It's probably cleaner if a reviewer does the merge.

**Comment by karussell** [comment]
> It's probably cleaner if a reviewer does the merge.

Good point!
